pkg/csi/service/wcp/controller.go

```diff
@@ -933,6 +1039,86 @@ func (c *controller) createFileVolume(ctx context.Context, req *csi.CreateVolume
 			VolumeContext: attributes,
 		},
 	}
+
+	// Calculate accessible topology for the provisioned volume in case of topology aware environment.
+	if isWorkloadDomainIsolationSupported {
+		if zoneLabelPresent {
+			selectedDatastore := volumeInfo.DatastoreURL
+			// CreateBlockVolumeUtil with idempotency enabled does not return datastore
+			// information when it uses the cached information from CR. In such cases,
+			// querying the volume to retrieve the datastore URL.
+			if selectedDatastore == "" {
+				queryFilter := cnstypes.CnsQueryFilter{
+					VolumeIds: []cnstypes.CnsVolumeId{
+						{
+							Id: volumeInfo.VolumeID.Id,
+						},
+					},
+				}
+				querySelection := cnstypes.CnsQuerySelection{
+					Names: []string{string(cnstypes.QuerySelectionNameTypeDataStoreUrl)},
+				}
+				queryResult, err := utils.QueryVolumeUtil(ctx, c.manager.VolumeManager, queryFilter, &querySelection,
+					true)
+				if err != nil || queryResult == nil || len(queryResult.Volumes) != 1 {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to find the datastore on which volume %q is provisioned. Error: %+v",
+						volumeInfo.VolumeID.Id, err)
+				}
+				selectedDatastore = queryResult.Volumes[0].DatastoreUrl
+			}
+
+			// Calculate accessible topology for the provisioned volume.
+			datastoreAccessibleTopology, err := c.topologyMgr.GetTopologyInfoFromNodes(ctx,
```

Comment:
We want to add all the zones the datastore has access to in the supervisor as node affinity terms, not only the ones mentioned in the topology requirement.

Reply:
This will be taken care by changes in https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/2985 once merged